website/blog/2024-08-29-autogen-pyfunc/index.md

```diff
@@ -0,0 +1,596 @@
+
+To leverage Model from Code, we must perform the following steps:
+
+1. Declare a [custom PyFunc](https://mlflow.org/docs/latest/traditional-ml/creating-custom-pyfunc/index.html)
+2. Leverage [mlflow.models.set_model](https://mlflow.org/docs/latest/python_api/mlflow.models.html?highlight=set_model#mlflow.models.set_model) to indicate which python object is our model.
+
+To achieve these steps, we simply copy the above and below code to a python file. For simplicity, you can just create a single Python file with both code snippets, but MLflow also supports specifying local dependencies when logging our model via the `code_paths` parameter in [mlflow.pyfunc.lod_model](https://mlflow.org/docs/latest/python_api/mlflow.pyfunc.html?highlight=pyfunc%20log_model#mlflow.pyfunc.log_model)
+
+**This step was omitted for brevity and must be done manually.**
+
+```python
+import mlflow
+
+class CatifyPyfunc(mlflow.pyfunc.PythonModel):
+  def predict(self, context, model_input, params):
+    import mlflow
+    mlflow.autogen.autolog()
+
+    creator = CatifyWithDalle(
+            name="creator",
+            max_consecutive_auto_reply=0,
+            system_message="Help me coordinate generating image",
+            llm_config={"model":"gpt-4"},
+        )
+
+    user_proxy = UserProxyAgent(name="User", human_input_mode="NEVER", max_consecutive_auto_reply=0, code_execution_config={
+                "work_dir": "output", # Location where code will be written
+                "use_docker": False # Use local jupyter execution environment instead of docker
+            })
+
+    return user_proxy.initiate_chat(
+        creator, message=model_input
+    )
+mlflow.models.set_model(CatifyPyfunc())
+```
+
+At the end of this step, you should have a Python file that has both code snippets. The name is up to the user, but for this blog we will use "catify_model.py".
+
+## 3 Use Our Agent Framework
+
+We are now positioned to leverage MLflow to interact with our powerful "catify" agent.
+
+### 3.1 - Log and Load
+
+First, let's demonstrate the standard user journey of logging model to MLflow's tracking server. We will then load it back and perform inference.
+
+```python
+import mlflow
+mlflow.autogen.autolog() # Enable logging of traces
+
+with mlflow.start_run() as run:
+    mlflow.pyfunc.log_model(
+        artifact_path="autogen_pyfunc",
+        python_model="catify_model.py", # Our model from code python file
+
```

Comment:
It might be worthwhile to log a JSON artifact to the run that has 2 prompts for 'testing' and using the `log_image` API for storing the resulting generated images just to showcase this functionality that exists within MLflow. The dict that is logged can be a mapping of the index of input to both the prompt and the logged image name. 
No need to re-generate the images for the examples here - they're great - just update the logging function here to iterate over a list of test prompts and construct those data structures / call the model to log the images.

Reply:
Autogen cache for the win! Great call